### PR TITLE
[SPARK-38505][SQL] Make partial aggregation adaptive

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

We can skip do partial aggregation to avoid spilling if this step does not reduce the number of rows too much. 

By setting `spark.sql.aggregate.adaptivePartialAggregationThreshold` to 0 this feature can be disabled.

### Why are the changes needed?

Improve partial aggregation phase performance and we can implement these 2 features after this PR:
1. SPARK-36245: Partial deduplicate the right side of left semi/anti join
2. SPARK-38506: Push partial aggregation through join

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test and TPC-H 5T benchmark test.

SQL | Before this PR(Seconds) | After this PR(Seconds)
-- | -- | --
q15 | 66  | 51
q17 | 86 | 80
q18 | 129 | 122

### CODE REVIEW

**Comment by wangyum** [comment]
```scala
import org.apache.spark.benchmark.Benchmark
val numRows = 1024 * 1024 * 50

spark.sql(s"CREATE TABLE t1 using parquet AS SELECT id  AS a, id % ${numRows / 10000} AS b, id % ${numRows / 10000} AS c, id AS d FROM range(1, ${numRows}L, 1, 10)")
val benchmark = new Benchmark("Benchmark WholeStageCodegenExec", numRows, minNumIters = 2)


Seq(0, 10000).foreach { threshold =>
  benchmark.addCase(s"SELECT a, c, sum(b), sum(d) FROM t1 where a > 100 group by a, c and partialAggThreshold=$threshold") { _ =>
    withSQLConf("spark.sql.aggregate.adaptivePartialAggregationThreshold" -> threshold.toString) {
      spark.sql("SELECT a, c, sum(b), sum(d) FROM t1 where a > 100 group by a, c").write.format("noop").mode("Overwrite").save()
    }
  }
}

benchmark.run()
```

```
Java HotSpot(TM) 64-Bit Server VM 1.8.0_281-b09 on Mac OS X 10.15.7
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
Benchmark WholeStageCodegenExec:                                                               Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
SELECT a, c, sum(b), sum(d) FROM t1 where a > 100 group by a, c and partialAggThreshold=0              56519          57012         697          0.9        1078.0       1.0X
SELECT a, c, sum(b), sum(d) FROM t1 where a > 100 group by a, c and partialAggThreshold=10000          41908          42369         653          1.3         799.3       1.3X

```

**Comment by wangyum** [comment]
Before this PR | After this PR
-- | --
image | image

**Comment by wangyum** [comment]
Before this PR | After this PR
-- | --
image | image

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -2907,6 +2907,18 @@ object SQLConf {
...
+    buildConf("spark.sql.aggregate.adaptivePartialAggregationThreshold")
+      .internal()
+      .doc("Minimum number of processed rows before partial aggregation can be skipped. " +
+        "By setting this value to 0 adaptive partial aggregation can be disabled.")
+      .version("3.3.0")
+      .intConf
+      .checkValue(threshold => threshold >= 0 && threshold < (1 << 16),
+        "The threshold value must be bigger than or equal to 0 and less than " +
+          s"1 << ${FAST_HASH_AGGREGATE_MAX_ROWS_CAPACITY_BIT.key}.")
+      .createWithDefault(10000)
```
Comment (YuzhouSun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:2920]: Just curious, any reason that the default value of adaptivePartialAggregationThreshold is 10,000? [In Trino the default is 100,000](https://github.com/trinodb/trino/blob/ab7bc3fa534110893b96232b6b1c6d36286dcbb6/core/trino-main/src/main/java/io/trino/sql/planner/OptimizerConfig.java#L86)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala [hunk 2]
```diff
@@ -146,10 +155,32 @@ case class HashAggregateExec(
...
+      conf.adaptivePartialAggregationThreshold < (1 << conf.fastHashAggregateRowMaxCapacityBit) && {
+      child
+        .collectUntil(p => p.isInstanceOf[WholeStageCodegenExec] ||
+          !p.isInstanceOf[CodegenSupport] ||
+          p.isInstanceOf[LeafExecNode]).forall {
+        case _: ProjectExec | _: FilterExec | _: ColumnarToRowExec => true
+        case _: SerializeFromObjectExec => true
+        case _: InputAdapter => true
+        // HashAggregateExec, ExpandExec, SortMergeJoinExec ...
+        case _ => false
```
Comment (YuzhouSun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala:170]: Why does it need to limit child node types? Is this for a performance reason? If so, after nodes like aggregate, join, and expand, skipping partial aggregate may still benefit performance. Why isn’t adaptivePartialAggregationThreshold good enough?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala [hunk 3]
```diff
@@ -847,18 +904,60 @@ case class HashAggregateExec(
...
+      val numberOfConsumedKeysTerm =
+        ctx.addMutableState(CodeGenerator.JAVA_LONG, "numberOfConsumedKeys")
+      val numAggSkippedRows = metricTerm(ctx, "numSkippedParAggRows")
+      val initExpr = declFunctions.flatMap(f => f.initialValues)
+      val emptyBufferKeyCode = GenerateUnsafeProjection.createCode(ctx, initExpr)
+
+      val numberOfKeys = if (fastHashMapTerm != null) {
+        s"$numberOfConsumedKeysTerm = $fastHashMapTerm.getNumKeys();"
+      } else if (hashMapTerm != null) {
+        s"$numberOfConsumedKeysTerm = $hashMapTerm.getNumKeys();"
```
Comment (YuzhouSun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala:917]: When fastHashMap is enabled, this implementation doesn’t count keys in regular hashMap. However, even when number of processed keys is less than fastHashMap capacity, the regularHashMap seems can still be non-empty when the fastHashMap
        1. unable to find group key within maxSteps (maxSteps = 2)
        2. run out of max supported page size. 

So even though adaptivePartialAggregationThreshold must be smaller than fastHashMap capacity, it's still possible that both fastHashMap and regular hashMap are non-empty. A potential fix can be `numberOfKeys = fastHashMap.getNumKeys() + hashMap.getNumKeys()`, and don't enable skipPartialAggregate once the aggregate spills. With this fix we can also remove [the adaptivePartialAggregationThreshold < fastHashMap's capacity check](https://github.com/apache/spark/pull/35806/files#diff-13c5b65678b327277c68d17910ae93629801af00117a0e3da007afd95b6c6764R2917-R2919)

**Comment by DenineLu** [comment]
I was interested in working on this, but I tested it with an online production task and found that the performance was regressing. Even though the aggregation time is shortened, the whole stage is more time consuming. Have you encountered this situation please?

**Comment by YuzhouSun** [comment]
> I was interested in working on this, but I tested it with an online production task and found that the performance was regressing. Even though the aggregation time is shortened, the whole stage is more time consuming. Have you encountered this situation please?

Hi @DenineLu, just curious, if possible, could you share more details about the regression case you encountered, please? For example, how much aggregation time is shortened, how much longer is the whole stage, number of partial aggregate input rows and number of output rows with and without the optimization, etc. Thank you.

BTW the author moved related changes to a newer and larger PR: https://github.com/apache/spark/pull/36552 (for [SPARK-38506](https://issues.apache.org/jira/browse/SPARK-38506)
Push partial aggregation through join)

**Comment by DenineLu** [comment]
Thanks for your reply, I have provided two graphs which show that the HashAgg node shortens the execution time, but the whole stage becomes slower. I guess maybe because of this code, HashAgg's time statistics are not accurate anymore.
```  
|if (!$initAgg) {
|  $initAgg = true;
|  $createFastHashMap
|  $addHookToCloseFastHashMap
|  $hashMapTerm = $thisPlan.createHashMap();
|  long $beforeAgg = System.nanoTime();
|  $doAggFuncName();
|  $aggTime.add((System.nanoTime() - $beforeAgg) / $NANOS_PER_MILLIS);
|  $shouldStopCheckCode
|}
|// This function cannot be counted in time, but will be executed frequently
|$childDoAgg
```

not_refine
The top picture does not add this patch, the bottom picture adds this patch.
refine

**Pull request closed without merging**